Properly handle speculationrules script element removal.

yoavweiss opened the pull request:
#### c843c2aa8e00388e2e0f86accedc3b6ea0f76e9f
<pre>
Properly handle speculationrules script element removal.
<a href="https://bugs.webkit.org/show_bug.cgi?id=303548">https://bugs.webkit.org/show_bug.cgi?id=303548</a>

Reviewed by Alex Christensen.

Currently when a speculationrules script element is removed, its parsed rules are still processed in the next microtask checkpoint.
This PR changes that and removeis the rules of such scripts, ensuring that no prefetch would happen if the removal happens within a microtask of the script&apos;s addition.

No new tests, but this progresses existing tests.

* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/mutate-rules.https-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/speculation-tags/deduped-and-sorted-tags.https_type=prefetch-expected.txt: Progression.
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::registerSpeculationRules): Pass an element.
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::processSpeculationRules): Pass a callback to process each rule.
* Source/WebCore/dom/LoadableSpeculationRules.cpp:
(WebCore::LoadableSpeculationRules::notifyFinished): Pass a nullptr instead of an element.
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::registerSpeculationRules): Pass an element.
(WebCore::ScriptElement::unregisterSpeculationRules): Remove the speculation rules associated with an element.
* Source/WebCore/dom/ScriptElement.h:
* Source/WebCore/dom/SpeculationRulesMatcher.cpp:
(WebCore::SpeculationRulesMatcher::hasMatchingRule): Pass a callback to process each rule.
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::removedFromAncestor): Call unregisterSpeculationRules.
* Source/WebCore/loader/DocumentPrefetcher.cpp:
(WebCore::DocumentPrefetcher::removePrefetch): Remove a prefetched URL from cache.
* Source/WebCore/loader/DocumentPrefetcher.h:
* Source/WebCore/loader/SpeculationRules.cpp:
(WebCore::SpeculationRules::parseSpeculationRules): Add rule to a container associated with its element.
(WebCore::SpeculationRules::unregisterSpeculationRules): Remove element-associated rules.
(WebCore::SpeculationRules::prefetchRules const): Deleted.
* Source/WebCore/loader/SpeculationRules.h:
(WebCore::SpeculationRules::prefetchRules const): Getter.

Canonical link: <a href="https://commits.webkit.org/304255@main">https://commits.webkit.org/304255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c7a5f41084a7bc1ad8367157909b13239d58c56d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134946 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7363 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142454 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86799 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136815 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7209 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103113 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70406 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137892 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5653 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120953 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83961 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5474 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3083 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3050 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114689 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39110 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145154 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7039 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39685 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111494 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7092 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5905 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111846 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28401 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5308 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117237 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60958 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7088 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35408 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6861 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70660 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7095 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6968 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/SpeculationRules.h:
```diff
@@ -100,14 +102,27 @@ class SpeculationRules : public RefCounted<SpeculationRules> {
     SpeculationRules() = default;
 
-    Vector<Rule> m_prefetchRules;
+    HashMap<Ref<Element>, Vector<Rule>> m_prefetchRulesByElement;
```
Can this be a WeakHashMap?  It doesn't need to prevent the elements from being destroyed.

yoavweiss replied:
done

achristensen07 commented on Source/WebCore/dom/SpeculationRulesMatcher.cpp:
```diff
@@ -104,23 +104,28 @@ static bool matches(const SpeculationRules::DocumentPredicate& predicate, Docume
         if (rule.predicate && matches(rule.predicate.value(), document, anchor))
-            return PrefetchRule { rule.tags, rule.referrerPolicy, rule.eagerness == SpeculationRules::Eagerness::Conservative || rule.eagerness == SpeculationRules::Eagerness::Moderate };
-    }
+            result = PrefetchRule { rule.tags, rule.referrerPolicy, rule.eagerness == SpeculationRules::Eagerness::Conservative || rule.eagerness == SpeculationRules::Eagerness::Moderate };
```
I do like the early return instead of the optional, capture, and quick iteration.  Would it be reasonable to keep track of prefetch rules by Node instead of by Element, then use the Document as the Node "owning" the rules from headers?

yoavweiss replied:
Done!

yoavweiss committed:
Properly handle speculationrules script element removal.
https://bugs.webkit.org/show_bug.cgi?id=303548

Reviewed by Alex Christensen.

Currently when a speculationrules script element is removed, its parsed rules are still processed in the next microtask checkpoint.
This PR changes that and removeis the rules of such scripts, ensuring that no prefetch would happen if the removal happens within a microtask of the script's addition.

No new tests, but this progresses existing tests.

* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/mutate-rules.https-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/speculation-tags/deduped-and-sorted-tags.https_type=prefetch-expected.txt: Progression.
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::registerSpeculationRules): Pass an element.
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::processSpeculationRules): Pass a callback to process each rule.
* Source/WebCore/dom/LoadableSpeculationRules.cpp:
(WebCore::LoadableSpeculationRules::notifyFinished): Pass a nullptr instead of an element.
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::registerSpeculationRules): Pass an element.
(WebCore::ScriptElement::unregisterSpeculationRules): Remove the speculation rules associated with an element.
* Source/WebCore/dom/ScriptElement.h:
* Source/WebCore/dom/SpeculationRulesMatcher.cpp:
(WebCore::SpeculationRulesMatcher::hasMatchingRule): Pass a callback to process each rule.
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::removedFromAncestor): Call unregisterSpeculationRules.
* Source/WebCore/loader/DocumentPrefetcher.cpp:
(WebCore::DocumentPrefetcher::removePrefetch): Remove a prefetched URL from cache.
* Source/WebCore/loader/DocumentPrefetcher.h:
* Source/WebCore/loader/SpeculationRules.cpp:
(WebCore::SpeculationRules::parseSpeculationRules): Add rule to a container associated with its element.
(WebCore::SpeculationRules::unregisterSpeculationRules): Remove element-associated rules.
(WebCore::SpeculationRules::prefetchRules const): Deleted.
* Source/WebCore/loader/SpeculationRules.h:
(WebCore::SpeculationRules::prefetchRules const): Getter.

Canonical link: https://commits.webkit.org/304255@main
